src/shred_network/duplicate_shred_listener.zig

```diff
@@ -0,0 +1,794 @@
+        self.params.duplicate_slots_sender.send(key.slot) catch |e| {
+            self.logger.err().logf(
+                "duplicate_shred_listener: send duplicate slot {} failed: {}",
+                .{ key.slot, e },
+            );
+        };
+
+        self.cleanupEntry(key);
+    }
+
+    fn cleanupEntry(self: *DuplicateShredHandler, key: Key) void {
+        if (self.dup_buffer.fetchRemove(key)) |kv| {
+            const entry = kv.value;
+            for (entry.chunks) |maybe_chunk| if (maybe_chunk) |chunk| self.allocator.free(chunk);
+        }
+    }
+
+    fn reconstructShredsFromData(
+        self: *DuplicateShredHandler,
+        key: Key,
+        data: []const u8,
+    ) !struct { sig.ledger.shred.Shred, sig.ledger.shred.Shred } {
+        const proof = sig.bincode.readFromSlice(
+            self.allocator,
+            DuplicateSlotProof,
+            data,
+            .{},
+        ) catch |e| {
+            self.logger.err().logf(
+                "duplicate_shred_listener: failed to deserialize proof for slot {}: {}",
+                .{ key.slot, e },
+            );
+            return error.InvalidDuplicateShreds;
+        };
+        defer sig.bincode.free(self.allocator, proof);
+
+        var shred1 = sig.ledger.shred.Shred.fromPayload(self.allocator, proof.shred1) catch {
+            return error.InvalidDuplicateShreds;
+        };
+        errdefer shred1.deinit();
+        var shred2 = sig.ledger.shred.Shred.fromPayload(self.allocator, proof.shred2) catch {
+            return error.InvalidDuplicateShreds;
+        };
+        errdefer shred2.deinit();
+
+        if (shred1.commonHeader().slot != key.slot or shred2.commonHeader().slot != key.slot) {
+            return error.SlotMismatch;
+        }
+
+        const sv: u16 = self.params.shred_version.load(.monotonic);
+        if (shred1.commonHeader().version != sv or shred2.commonHeader().version != sv) {
+            return error.InvalidShredVersion;
+        }
+
+        const leader = self.params.leader_schedule.get(key.slot) orelse {
+            return error.UnknownLeader;
+        };
+        shred1.verify(leader) catch {
+            return error.InvalidSignature;
+        };
+        shred2.verify(leader) catch {
+            return error.InvalidSignature;
+        };
+
+        const same_fec =
+            shred1.commonHeader().erasure_set_index == shred2.commonHeader().erasure_set_index;
+        const mr1 = shred1.merkleRoot() catch null;
+        const mr2 = shred2.merkleRoot() catch null;
+        var conflict_ok = false;
+        if (same_fec) {
+            if ((mr1 == null and mr2 != null) or (mr1 != null and mr2 == null)) {
+                conflict_ok = true;
+            } else if (mr1) |h1| {
+                if (mr2) |h2| {
+                    if (!std.mem.eql(u8, &h1.data, &h2.data)) {
+                        conflict_ok = true;
+                    }
+                }
+            }
```

Comment:
Since their non-nullness is confirmed to be in the same state at this point:
```suggestion
            } else if (mr1) |h1| {
                const h2 = mr2.?;
                if (!std.mem.eql(u8, &h1.data, &h2.data)) {
                    conflict_ok = true;
                }
            }
```

Reply:
[eb9ca4a](https://github.com/Syndica/sig/pull/1027/commits/eb9ca4a660a2e0b3a3bbd007ea705a26d86d3b64)